pkg/provider/azure_vmss.go

```diff
@@ -2359,3 +2353,89 @@ func (ss *ScaleSet) VMSSBatchSize(ctx context.Context, vmssName string) (int, er
 	klog.V(2).InfoS("Fetch VMSS batch size", "vmss", vmssName, "size", batchSize)
 	return batchSize, nil
 }
+
+func UpdateVMSSVMsInBatch(ctx context.Context, update map[string]armcompute.VirtualMachineScaleSetVM, batchSize int, patchVMFn func(ctx context.Context, instanceID string, vm *armcompute.VirtualMachineScaleSetVM) (*runtime.Poller[armcompute.VirtualMachineScaleSetVMsClientUpdateResponse], error)) <-chan error {
+	logger := klog.FromContext(ctx)
+	if patchVMFn == nil {
+		logger.Error(fmt.Errorf("patchVMFn is nil"), "Failed to update VMs for VMSS with new backendPoolID")
+		return nil
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	errChan := make(chan error, len(update))
+	pollerChannel := make(chan *runtime.Poller[armcompute.VirtualMachineScaleSetVMsClientUpdateResponse], len(update))
+	var pollerGroup sync.WaitGroup
+	pollerGroup.Add(1)
+	go func() {
+		defer pollerGroup.Done()
+		for {
+			select {
+			case poller, ok := <-pollerChannel:
+				if !ok {
+					// pollerChannel is closed
+					return
+				}
+				if poller == nil {
+					continue
+				}
+				pollerGroup.Add(1)
+				go func() {
+					defer pollerGroup.Done()
+					_, err := poller.PollUntilDone(ctx, &runtime.PollUntilDoneOptions{
+						Frequency: 10 * time.Second,
+					})
+					if err != nil {
+						logger.Error(err, "Failed to update VMs for VMSS with new vm config")
+						errChan <- err
+					}
+				}()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	logger.Info("Begin to update VMs for VMSS with new vm config")
+	if batchSize > 1 {
+		concurrentLock := make(chan struct{}, batchSize)
+		var requestGroup sync.WaitGroup
+		for instanceID, vm := range update {
+			instanceID := instanceID
+			vm := vm
+			concurrentLock <- struct{}{}
+			requestGroup.Add(1)
+			go func() {
+				defer func() {
+					requestGroup.Done()
+					<-concurrentLock
+				}()
+				poller, rerr := patchVMFn(ctx, instanceID, &vm)
+				if rerr != nil {
+					errChan <- rerr
+					return
+				}
+				if poller != nil {
+					pollerChannel <- poller
+				}
+			}()
+		}
+		requestGroup.Wait()
+		close(concurrentLock)
+	} else {
```

Comment:
Can we merge the if blocks? It would be equivalent to create 1 goroutine and wait.

Reply:
Agree, it brings less maintenance effort.

Reply:
It will eliminate context switch in go. Considering the fact that batch size is 1 in most cases, it might be worth maintaining two blocks. Because once a change is made, it seldomly changes.

Reply:
BTW, merging the block is what we did previously.
I guess Shangxiang's concern is AKS uses batch 1 as default, which means we'll always start goroutines to update vm in a series, which seems to be unnecessary?